lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -170,32 +199,41 @@ performSelection
     :: forall m. (HasCallStack, MonadRandom m)
     => SelectionCriteria
     -> m (Either SelectionError SelectionResult)
-performSelection SelectionCriteria {outputsToCover, utxoAvailable}
+performSelection criteria
     | not (balanceRequired `leq` balanceAvailable) =
         pure $ Left $ BalanceInsufficient $ BalanceInsufficientError
             {balanceAvailable, balanceRequired}
     | otherwise =
-        Right . mkResult <$> runSelection utxoAvailable balanceRequired
+        mkResult <$>
+            runSelection selectionLimit utxoAvailable balanceRequired
   where
+    SelectionCriteria
+        {outputsToCover, utxoAvailable, selectionLimit} = criteria
+
     balanceAvailable :: TokenBundle
     balanceAvailable = view #balance utxoAvailable
 
     balanceRequired :: TokenBundle
     balanceRequired = F.foldMap (view #tokens) outputsToCover
 
-    mkResult :: SelectionState -> SelectionResult
-    mkResult SelectionState {selected, leftover} =
-        case NE.nonEmpty (UTxOIndex.toList selected) of
-            Nothing ->
-                unableToSelectAnyInputsError
-            Just inputsSelected ->
-                SelectionResult
+    mkResult :: SelectionState -> Either SelectionError SelectionResult
+    mkResult SelectionState {selected, leftover}
+        | not (balanceRequired `leq` balanceSelected) =
+            Left $ SelectionInsufficient $ SelectionInsufficientError
+                {inputsSelected = UTxOIndex.toList selected, balanceRequired}
```

Comment:
I like this approach. We could have callers to performSelection possibly retry on `SelectionInsufficient` errors. If we're close to the required balance, the selection may still succeed with a different selection of inputs.

Reply:
> We could have callers to performSelection possibly retry on SelectionInsufficient errors. If we're close to the required balance, the selection may still succeed with a different selection of inputs.

I think that's a great idea. 👍🏻